### PR TITLE
Refactor and fix per-age-class history outputs

### LINKED ISSUE #1205: Incorrect (?) calculation of per-age-class outputs
There are (at least) two different ways that per-age-class (`_AP`) outputs are be calculated in `update_history_dyn()`.

When I look at such a variable, I expect the value in each age class to be the mean across all patches in that age class, weighted by patch area. This is how it's done for LAI (`FATES_LAI_AP`), requiring two steps. First, in `patchloop`:
```fortran
hio_lai_si_age(io_si,cpatch%age_class) = hio_lai_si_age(io_si,cpatch%age_class) &
   + sum(cpatch%tlai_profile(:,:,:)) * cpatch%area
```

Then you need to divide by the total age class area. Since that's accumulated during `patchloop`, this happens after that loop:

```fortran
! divide so-far-just-summed but to-be-averaged patch-age-class
! variables by patch-age-class area to get mean values
do ipa2 = 1, nlevage
   if (hio_area_si_age(io_si, ipa2) .gt. nearzero) then
         hio_lai_si_age(io_si, ipa2) = hio_lai_si_age(io_si, ipa2) / (hio_area_si_age(io_si, ipa2)*AREA)
         ...
   else
      hio_lai_si_age(io_si, ipa2) = 0._r8
      ...
   endif
end do
```

However, for `FATES_FUEL_AMOUNT_AP`, it's all done in one step within `patchloop`:
```fortran
 ! Fuel sum [kg/m2]
 hio_fire_sum_fuel_si_age(io_si, cpatch%age_class) = hio_fire_sum_fuel_si_age(io_si, cpatch%age_class) +  &
    cpatch%sum_fuel * cpatch%area * AREA_INV
```
This doesn't give the average fuel load in each patch age class. Instead, it gives each age class's *contribution* to the total fuel load. So if you do an unweighted sum across all age classes, you'll get the total fuel load:
screenshot_0862

As expected, this isn't the case for LAI:
screenshot_0864

I don't know if this is intended behavior for `FATES_LAI_AP`, but as I mentioned it's certainly not what I would expect as a user of the output. Of outputs with `fates_levage` dimension (I haven't checked outputs with age class crossed by something else, like PFT), the following outputs (in addition to `FATES_FUEL_AMOUNT_AP`) appear to follow the same behavior:
- `FATES_ZSTAR_AP`
- `FATES_SECONDAREA_ANTHRODIST_AP`
- `FATES_SECONDAREA_DIST_AP`
- `FATES_BURNFRAC_AP`
- `FATES_CANOPYAREA_AP`
- `FATES_VEGC_AP`

### PR DESCRIPTION
**Pull request by samsrabin**
### Description:
Will resolve #1205.

### Collaborators:
@glemieux @rgknox @ckoven 

### Checklist

*If this is your first time contributing, please read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/main/CONTRIBUTING.md) document.*

All checklist items must be checked to enable merging this pull request:

*Contributor*
- [ ] The in-code documentation has been updated with descriptive comments
- [ ] The documentation has been assessed to determine if updates are necessary

*Integrator*
- [ ] FATES PASS/FAIL regression tests were run
- [ ] Evaluation of test results for answer changes was performed and results provided

### Documentation

- [Technical Note](https://github.com/NGEET/fates-docs) update:
- [User's Guide](https://github.com/NGEET/fates-users-guide) update: 

### Test Results:

*CTSM (or) E3SM (specify which) test hash-tag:*

*CTSM (or) E3SM (specify which) baseline hash-tag:*

*FATES baseline hash-tag:*

*Test Output:*

### CODE REVIEW

File: main/FatesHistoryInterfaceMod.F90 [hunk 1]
```diff
@@ -3393,46 +3398,58 @@ subroutine update_history_dyn2(this,nc,nsites,sites,bc_in)
 
 
                 cpatch%age_class  = get_age_class_index(cpatch%age)
+                age_class_area = sites(s)%area_by_age(cpatch%age_class)
 
                 ! Increment the fractional area in each age class bin
-                hio_area_si_age(io_si,cpatch%age_class) = hio_area_si_age(io_si,cpatch%age_class) &
+                hio_fracarea_si_age(io_si,cpatch%age_class) = hio_fracarea_si_age(io_si,cpatch%age_class) &
```
Comment (samsrabin) [inline main/FatesHistoryInterfaceMod.F90:3404]: This and the equivalent hio_fracarea_si_landuse code below can now actually be moved out of the patch loop and into an age-class loop.

**Comment by samsrabin** [comment]
As of `8e7a1d85`, diffs are as expected: only in `_AP` variables, and root mean square differences all < 1e-15.

**Comment by samsrabin** [comment]
Unfortunately, my test script indicates that most variables are still not getting the right answer. Here, ✅ indicates that ageclass-area-weighted mean of the per-ageclass (`_AP`) variable matches the non-per-ageclass variable, and ❌ indicates otherwise. The first symbol indicates the result from before I started making changes, and the second is from commit 1ec6d6eb.

```
❌ → ✅ FATES_BURNFRAC(_AP):
     max abs diff = 5.9e-08 → 8.99e-10
     max rel diff = 99.9% → 52.5%
❌ → ❌ FATES_DDBH_CANOPY_SZ(AP):
     max abs diff = 0.0262 → 0.000369
     max rel diff = 99.1% → 13.0%
❌ → ❌ FATES_DDBH_USTORY_SZ(AP):
     max abs diff = 0.0221 → 0.000118
     max rel diff = 96.2% → 2.3%
❌ → ❌ FATES_FIRE_INTENSITY_BURNFRAC(_AP):
     max abs diff = 2.62e+04 → 323
     max rel diff = 99.9% → 52.5%
❌ → ❌ FATES_FUEL_AMOUNT(_AP):
     max abs diff = 0.506 → 0.0108
     max rel diff = 59.3% → 5.5%
❌ → ❌ FATES_FUEL_AMOUNT_(AP)FC:
     max abs diff = 0.354 → 0.00708
     max rel diff = 96.1% → 7.9%
✅ → ✅ FATES_GPP(_AP)
❌ → ❌ FATES_LAI(_AP):
     max abs diff = 0.00524
     max rel diff = 7.7%
❌ → ❌ FATES_LBLAYER_COND(_AP):
     max abs diff = 0.0962
     max rel diff = 8.5%
❌ → ❌ FATES_MORTALITY_CANOPY_SZ(AP):
     max abs diff = 17.5 → 17.5
     max rel diff = 99.1% → 116.7%
❌ → ❌ FATES_MORTALITY_USTORY_SZ(AP):
     max abs diff = 2.63e+03 → 2.63e+03
     max rel diff = 100.0% → 100.0%
❌ → ❌ FATES_NPLANT_CANOPY_SZ(AP):
     max abs diff = 30.4 → 2.15
     max rel diff = 99.1% → 7.3%
❌ → ❌ FATES_NPLANT_SZ(AP):
     max abs diff = 67.3 → 2.15
     max rel diff = 99.1% → 7.3%
❌ → ❌ FATES_NPLANT_USTORY_SZ(AP):
     max abs diff = 58.3 → 0.996
     max rel diff = 96.2% → 3.8%
✅ → ✅ FATES_NPP(_AP)
❌ → ✅ FATES_NPP_(AP)PF:
     max abs diff = 6.06e-08 → 2.18e-10
     max rel diff = 247.0% → 65.9%
❌ → ❌ FATES_STOMATAL_COND(_AP):
     max abs diff = 0.00179
     max rel diff = 41.4%
❌ → ❌ FATES_VEGC(_AP):
     max abs diff = 0.446 → 0.0118
     max rel diff = 58.6% → 7.4%
❌ → ❌ FATES_VEGC_(AP)PF:
     max abs diff = 0.312 → 0.00849
     max rel diff = 100.0% → 100.0%

🤷 Non-per-age equivalent not in Dataset:
     FATES_CANOPYAREA(_AP)
     FATES_NCL(_AP)
     FATES_NPATCH(_AP)
     FATES_PATCHAREA(_AP)
     FATES_SCORCH_HEIGHT_(AP)PF
     FATES_SECONDAREA_ANTHRODIST(_AP)
     FATES_SECONDAREA_DIST(_AP)
     FATES_ZSTAR(_AP)

🤷 Too many (> 2) duplexed dimensions:
     FATES_NPLANT_SZ(AP)PF
```

This is from a 4-year, 10x15 run with CTSM on Derecho.

**Comment by samsrabin** [comment]
Looks like `isclose()` might just be too strict. [Plotting the discrepancies as box plots](https://peach-ebonee-47.tiiny.site/), it looks like most are at least drastically improved. I think this PR would thus be an improvement, even if it doesn't end up being perfect.

I'll have a look at the unchanged variables before submitting.

**Comment by samsrabin** [comment]
A note: I've updated my analysis script to use `FATES_CANOPYAREA_AP` as the weighting for `FATES_STOMATAL_COND_AP` and `FATES_LBLAYER_COND_AP`. However, because `FATES_CANOPYAREA_AP` has also been fixed as part of this PR, it can't be used straight-up as a weight. Instead, after 5942a0d, it must be multiplied by `FATES_PATCHAREA_AP`. This can result in a slightly worse discrepancy between the `_AP` and non-`_AP` variables, since there's more opportunity for imprecision to slip in.

**Comment by samsrabin** [comment]
Alright, most everything is improved now; see the [latest analysis outputs](https://samsrabin.github.io/analysis-outputs/fates-refactor-history/tests_0718-095838de.tests_0724-125943de.SMS_Lm49.f10_f10_mg37.I2000Clm60Fates.derecho_intel.clm-FatesColdAllVarsMonthly.html) comparing before I did anything vs. f21fa95.

Incremental changes:
- [`ff87ce15` to `103fdc96`](https://samsrabin.github.io/analysis-outputs/fates-refactor-history/tests_0718-095838de.tests_0723-141100de.SMS_Lm49.f10_f10_mg37.I2000Clm60Fates.derecho_intel.clm-FatesColdAllVarsMonthly.html)
- [`a0881c536`](https://samsrabin.github.io/analysis-outputs/fates-refactor-history/tests_0723-141100de.tests_0724-101913de.SMS_Lm49.f10_f10_mg37.I2000Clm60Fates.derecho_intel.clm-FatesColdAllVarsMonthly.html)
- [`f21fa95b`](https://samsrabin.github.io/analysis-outputs/fates-refactor-history/tests_0724-101913de.tests_0724-125943de.SMS_Lm49.f10_f10_mg37.I2000Clm60Fates.derecho_intel.clm-FatesColdAllVarsMonthly.html): Here, `FATES_MORTALITY_CANOPY_SZAP` got WORSE for some reason, although it's still much better than before I made any changes.

I think this is ready to have someone look at it. Any suggestions as to what I'm doing wrong with the mortality variables would be much appreciated, but it'd be good to at least get opinions on the other stuff.

**Comment by glemieux** [comment]
@rgknox @ckoven and @samsrabin to schedule a meeting to review.

**Comment by samsrabin** [comment]
Superseded by PR #1252.

**Pull request closed without merging**